### PR TITLE
New update reaction inputs

### PR DESCRIPTION
**Pull request by TorkelE**
https://github.com/SciML/Catalyst.jl/pull/856

### CODE REVIEW

File: src/reaction.jl [hunk 1]
```diff
@@ -160,83 +164,74 @@ struct Reaction{S, T}
...
-        stoich_type = Any
-        substoich′ = Any[value(s) for s in substoich]
-        prodstoich′ = Any[value(p) for p in prodstoich]
-    else
-        substoich′ = (S == stoich_type) ? substoich : convert.(stoich_type, substoich)
-        prodstoich′ = (T == stoich_type) ? prodstoich : convert.(stoich_type, prodstoich)
-    end
+    (stoich_type <: Num) && (stoich_type = Any)
+    substoich = promote_reaction_vector(substoich, stoich_type)
+    prodstoich = promote_reaction_vector(prodstoich, stoich_type)
```
Comment (isaacsas) [inline src/reaction.jl:188]: This doesn't seem functionally equivalent to the old code? Why this change? (It is specifically missing the `convert` calls?)

**Review by isaacsas — CHANGES_REQUESTED** [review]
I'm going to stop with the detailed comments -- I haven't read through all of this.

I don't see the reason for all the typing / conversion changes being put in place here. If there are gains in performance, allocations, or we are adding new functionality please let me know and I can look them over. If these are just stylistic because you find the code you've written easier to understand I'd prefer if we leave this code alone since it has worked just fine for a long time and not had any issues that I am aware of. 

If there aren't any substantiative reasons for such changes, then I'd prefer this to just focus on generalizing the allowed null inputs to the constructor, and leave the rest of the type conversion / promotion code and netstoich code alone at this time. I am happy to revisit updating this should we have benchmarks showing improved performance or reduced allocations with a different approach.

File: test/reactionsystem_core/reaction.jl [hunk 2]
```diff
@@ -36,13 +36,94 @@ let
...
+    @species X(t) Y(t) Z(t)
+    @variables A(t)
+
+    # Creates `Reaction`s.
+    rx1 = Reaction(k*A, [X], [])
+    rx2 = Reaction(k*A, [x], [Y], [1.5], [1])
+    rx3 = Reaction(k*A, [x, X], [], [n1 + n2, n2], [])
+    rx4 = Reaction(k*A, [X, Y], [X, Y, Z], [2//3, 3], [1//3, 1, 2])
+    rx5 = Reaction(k*A, [X, Y], [X, Y, Z], [2, 3], [1, n1, n2])
+    rx6 = Reaction(k*A, [X], [x], [n1], [1])
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply test/reactionsystem_core/reaction.jl:58]: Yes it definitely does.

File: src/reaction.jl [hunk 3]
```diff
@@ -134,19 +138,19 @@ Notes:
...
+    products::Vector{Any}
     """The stoichiometric coefficients of the reactants."""
-    substoich::Vector{T}
+    substoich::Vector
     """The stoichiometric coefficients of the products."""
-    prodstoich::Vector{T}
+    prodstoich::Vector
     """The net stoichiometric coefficients of all species changed by the reaction."""
-    netstoich::Vector{Pair{S, T}}
+    netstoich::Vector{Pair{Any, T}}
```
[the remark this answers is not in the document]
Comment (TorkelE) [reply src/reaction.jl:153]: When I first started looking at this I realised that we didn't typed everything as stringently as we could (i.e. first I figured `substrates::Vector{R}` would be better than just `substrates::Vector`, and then I realised we actually knew the type so could insert it. I think I was still in the old mode when we just narrowed down things as much as possible. I agree that if there is any doubt, there is no need. I really am not familiar with the impact of typing, I pretty much started doing it because you where doing it, but not based on any actual insight. If you have any doubts about any typing I 100% leave it to you to make that decision.

I also properly messed-up when I tried to revert to use the old typing system. That was not my intention but a really bad mistake, I am sorry about that.

**Comment by TorkelE** [comment]
Sorry for messing up the typing and reverting, it should be all sorted now. I realise at this stage there are way more change than what is actually needed for sorting out the argument only, if you want I could make a new PR tomorrow that basically just appends the arguments to the old `Reaction` constructor, that should be fairly quick (and just copy the added tests over as well).

**Comment by isaacsas** [comment]
I'd prefer to just merge a PR that addresses only the more flexible inputs and leaves the rest alone. We understand and are happy with the performance characteristics of the current code, and without benchmarking / allocation checking I don't understand how these changes might impact things. Given creating `Reaction`s is fundamental to model construction I'd prefer if we just keep this as is in all the typing-related code until we can more systematically check about performance/allocations as we make changes. (I'm not against updating it, but I am against modifying it without extensive benchmarking given it works just fine now and we aren't having major performance issues.)

**Comment by TorkelE** [comment]
sounds good, new version is in https://github.com/SciML/Catalyst.jl/pull/918
